### DISCUSSION TITLE
Slower sources with higher priorities causes flashing/flickering. Any workaround?

### DISCUSSION THREAD

**Discussion by HymanZHAN** [hrsh7th/nvim-cmp#1280] [Q&A] [answered]
In my `cmp` config, I set source `nvim_lsp` with priority 1000 and `luasnip` with priority 750, and that causes some undesirable flashing in the autocompletion dropdown. 

![cmp with rust-analyzer](https://user-images.githubusercontent.com/26806995/199650245-f2ffba4c-703a-4947-8533-f04fe7f0b8fc.gif)

I mainly work with Python, Rust, and JS/TS, whose LSP implementations are pretty mainstream but would still require around 120 to 160ms for the initial round trip (on my machine). On the other hand, `luasnip` is pretty instantaneous, so it would always appear first, then get preceded by LSP options.

Is there a way to work around this? Something like waiting for all sources to complete and then showing the autocompletion dropdown? I wouldn't mind a 200ms delay for auto-completion when I type.

**Comment by HymanZHAN** [comment]
BTW, I have tried `performance.debounce` but that's not the solution, unfortunately.

**Comment by Shougo** [comment]
Hm...  It is from source asynchronous update.
It seems `nvim-cmp` feature.

**Comment by hrsh7th** [comment]
OK. I'll add an option `performance.source_timeout`

**Reply by hrsh7th** [reply]
Ah. Sorry. It's already implemented.

You can change `performance.fetching_timeout`

**Reply by HymanZHAN** [reply]
Wow that's quick (and beyond expectation 😂). Many thanks!

**Reply by HymanZHAN** [reply]
Thanks for pointing this option out. I will give it a try!

**Comment by HymanZHAN** [comment]
Ok I experimented with `performance.fetching_timeout` of 600, but the issue persists. I will try again with a minimal config.

**Reply by HymanZHAN** [reply]
Weird enough, I am not able to reproduce this behavior with my own config. :man_facepalming: The problem I reported originally was found on a nvim config distribution called [Astronvim](https://astronvim.github.io/).

With my own config, I have:

[Screencast from 11-05-2022 01:46:28 AM.webm](https://user-images.githubusercontent.com/26806995/200042008-d67ddd32-ef83-49ed-bad2-6b7771155bac.webm)

The autocompletion dropdown appears with all sources loaded. No async insertion occurred. I think this is as expected because the default value of `performance.fetching_timeout` is 200ms.

I am unsure if it's because I am loading snippets from `friend-snippets` eagerly in packer (without `opt=true` in packer) in my own config. I have [another discussion here](https://github.com/AstroNvim/AstroNvim/issues/1248) with the Astronvim author with some more details, but I will play with my Astronvim config and see if anything comes up.

**Reply by HymanZHAN** [reply]
My personal neovim config: https://gitee.com/zhanxucong/neovim_config
My Astronvim user config: https://gitee.com/zhanxucong/astrovim_config
In case you are interested :)

**Comment by HymanZHAN** [comment, marked as answer]
Ah, I found the root cause. I was using a custom comparator in `sorting.comparators` that was calling `vim.api.nvim_buf_get_lines()`, which is async. I had been aware of this but after I removed this method call, the async inserting behavior persisted. As it turned out, me being a noob totally forgot to run `:PackerCompile` after updating my `init.lua`. :man_facepalming: Sorry for all the hassle and thanks for the support! :sweat_smile: :hugs:
